### PR TITLE
Use a default Java version from root POM

### LINKED ISSUE #30275: If docs/building.md is serious about the Java version requirement, set it as `<maven.compiler.*>` to not allow lower ones
### Description

Quoth `docs/building.md`:

```
Ensure you have JDK 21 (or newer) and Git installed

    java -version
    git --version
```

If this is true, then the build system should not allow a version of Java less than 21. I'm filing this because it should be tracked as something that needs doing. I do notice that Java 21 support was implemented 5 days ago in 262fc09edc5, but nevertheless.

### PR DESCRIPTION
**Pull request by jonkoops**
Sets all the Maven compiler specific properties to Java 17, unless older compiler targets are needed for specific POMs, in which case they are overwritten selectively to an older version. This allows us to assume a base version of 17 everywhere in the project, rather than just in specific areas where this version was previously overwritten selectively.

This also upgrades Maven to the latest version, as well as some relevant plugins such as the [Maven Plugin Plugin](https://maven.apache.org/plugin-tools/maven-plugin-plugin/), and ensures that only a single version of these dependencies is used in the entire project.

Additionally, any profiles that were previously only triggered on specific JDK versions that have a version of 17 or higher have now been in-lined, as we can now safely assume this is the minimum version we support.

Closes #30275

### CODE REVIEW

**Review by stianst — CHANGES_REQUESTED** [review]
Only the server has a minimum of Java 17, remaining adapters (and shared libraries) still need to support older Java versions.

@mposolda FIY

**Comment by jonkoops** [comment]
> Only the server has a minimum of Java 17, remaining adapters (and shared libraries) still need to support older Java versions.

Yeah, I'll keep those on the correct version. The main goal for this PR is to set the default to 17, and selectively overwrite it  to a lower version where needed, rather than selectively opting into 17.

File: pom.xml
Comment (mabartos) [inline pom.xml:1735]: IMO, the profile might be removed, and the `surefire.system.args` props set by default. As it should touch only the tests exec and we do not test even the JBoss EAP SAML adapter on JDK < 9, we should be ok.

File: pom.xml
Comment (mabartos) [inline pom.xml:1839]: I'd opt to remove the profile and include the plugin to the default plugins.

**Review by mabartos — COMMENTED** [review]
> The main goal for this PR is to set the default to 17, and selectively overwrite it to a lower version where needed, rather than selectively opting into 17.

@jonkoops Agree with the statement. Added a few comments.

File: adapters/spi/jboss-adapter-core/pom.xml
Comment (jonkoops) [inline adapters/spi/jboss-adapter-core/pom.xml:34]: Same here, should we log an issue to make this compile with a more modern version instead?

File: pom.xml
Comment (jonkoops) [reply pom.xml]: Agreed, I've moved this to be the default and linked the original issue in a comment for future reference.

File: pom.xml
Comment (jonkoops) [reply pom.xml]: Agreed, I've moved the plugin.

File: testsuite/db-allocator-plugin/pom.xml
Comment (jonkoops) [inline testsuite/db-allocator-plugin/pom.xml:74]: Using a newer version of `maven-plugin-plugin` requires us to provide a `goalPrefix` where this was previously inferred. I have specified this based on what is documented under the ["Plugin Prefix Resolution"](https://maven.apache.org/guides/introduction/introduction-to-plugin-prefix-mapping.html) documentation, but I'd like another set of eyes to confirm this is correct.

Incidentally, it looks like we can omit this if we were following the naming [conventions for plugins](https://maven.apache.org/guides/introduction/introduction-to-plugin-prefix-mapping.html#specifying-a-plugins-prefix), which would rename the plugin to `db-allocator-maven-plugin`.

File: adapters/saml/core/pom.xml
Comment (mabartos) [reply adapters/saml/core/pom.xml:36]: @jonkoops AFAIK, we still need to support the JBoss EAP SAML adapter, and I don't know the exact requirements in terms of JDKs. 

Maybe @pskopek or @miquelsi know some details? 

btw. What do you mean by "we're using removed APIs" ?

File: adapters/spi/jboss-adapter-core/pom.xml
Comment (mabartos) [reply adapters/spi/jboss-adapter-core/pom.xml:34]: @jonkoops Still not sure about the EAP SAML adapter; needs to be verified.

File: core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java
Comment (mabartos) [inline core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java:31]: It does not belong here. 

It seems some changes slipped here unintentionally. @jonkoops Correct?

File: core/src/main/java/org/keycloak/jose/jwk/JWKParser.java
Comment (mabartos) [inline core/src/main/java/org/keycloak/jose/jwk/JWKParser.java:35]: Same as above

File: docs/documentation/tests/pom.xml
Comment (mabartos) [inline docs/documentation/tests/pom.xml:11]: It should stay here, right? As we support JDK 17 and JDK 21. Or can you verify the warning is not present when using specific version of JDK 17?

As the properties are different than in https://github.com/keycloak/keycloak/pull/29927/files#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8R215

File: testsuite/db-allocator-plugin/pom.xml
Comment (mabartos) [reply testsuite/db-allocator-plugin/pom.xml:74]: If it's the case only for the db-allocator, I would not bother about it as db-allocator will be removed soon. See https://github.com/keycloak/keycloak/pull/30406

**Review by mabartos — CHANGES_REQUESTED** [review]
@jonkoops nice job! Just a few things.

File: adapters/saml/core/pom.xml
Comment (jonkoops) [reply adapters/saml/core/pom.xml:36]: The SAML adapter makes use of [`java.security.acl.Group`](https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/security/acl/Group.html), which was removed in JDK 14.

File: adapters/spi/jboss-adapter-core/pom.xml
Comment (jonkoops) [reply adapters/spi/jboss-adapter-core/pom.xml:34]: Just FYI the reason why this one needs to be down-leveled is because the SAML adapter makes use of [`java.security.acl.Group`](https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/security/acl/Group.html), which was removed in JDK 14.

File: testsuite/db-allocator-plugin/pom.xml
Comment (jonkoops) [reply testsuite/db-allocator-plugin/pom.xml]: I see now that the PR you mention was merged and this file no longer exists, so we can consider this resolved.

File: core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java
Comment (jonkoops) [reply core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java:31]: These are actually intentional changes, this code was previously in a separate directory that was activated by [a profile](https://github.com/keycloak/keycloak/blob/main/core/pom.xml#L78-L115) when the JDK version was 16 or higher. So essentially depending on the JDK version that you are using during build you would end up with the code from `src/main/java16` or `src/main/java`.

This change moves the code previously in the `src/main/java16` directory directly into `src/main/java`, as we can now always assume that we're compiling for JDK 17 or higher.

File: core/src/main/java/org/keycloak/jose/jwk/JWKParser.java
Comment (jonkoops) [reply core/src/main/java/org/keycloak/jose/jwk/JWKParser.java:35]: See https://github.com/keycloak/keycloak/pull/29927#discussion_r1638430175

File: docs/documentation/tests/pom.xml
Comment (jonkoops) [reply docs/documentation/tests/pom.xml:11]: You are correct, I thought they were the same properties, but I misread. I've put this back.

File: adapters/saml/core/pom.xml
Comment (mabartos) [reply adapters/saml/core/pom.xml:36]: I see, thanks for the explanation.

File: core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java
Comment (mabartos) [reply core/src/main/java/org/keycloak/jose/jwk/JWKBuilder.java:31]: @jonkoops Interesting. Ok, that works for me.

File: core/src/main/java16/org/keycloak/jose/jwk/JWKBuilder.java
Comment (mabartos) [inline core/src/main/java16/org/keycloak/jose/jwk/JWKBuilder.java:1]: @jonkoops So, even the `java16` dir is removed?

File: core/src/main/java/org/keycloak/jose/jwk/JWKParser.java
Comment (jonkoops) [reply core/src/main/java/org/keycloak/jose/jwk/JWKParser.java:1]: Oh interesting! I just moved the file in IntelliJ, so perhaps the original file had these permissions set.

File: core/src/main/java16/org/keycloak/jose/jwk/JWKBuilder.java
Comment (jonkoops) [reply core/src/main/java16/org/keycloak/jose/jwk/JWKBuilder.java:1]: Yes, these files have been copied into the `java` directory and the profile that would activate this code path (depending on the JDK version used at compile time) has been removed. This is the code I was referring to under https://github.com/keycloak/keycloak/pull/29927#discussion_r1638430175.

**Comment by jonkoops** [comment]
@tnorimat since you wrote the JWK code that I moved around here (plus the `java16` profile), could I ask you to vet these changes?

**Review by tnorimat — APPROVED** [review]
@jonkoops LGTM. Thank you.

File: adapters/saml/core/pom.xml
Comment (mposolda) [reply adapters/saml/core/pom.xml:36]: Regarding SAML adapter: I think that we may support it just with Wildfly 28 and EAP 8 and hopefully we will be able to support only "jakarta" and remove JEE/javax based artifacts. Some related GH issues are https://github.com/keycloak/keycloak/issues/30324 and https://github.com/keycloak/keycloak/issues/29922 . So once JEE SAML adapters are removed, we can hopefully stick to the lowest java version supported by Wildfly 28 / EAP 8.

But for now, we can hopefully keep as it is in this PR? And then possibly adjust as a follow-up issue based on Wildfly 28 / EAP 8.

**Review by mabartos — DISMISSED** [review]
@jonkoops LGTM

**Comment by mposolda** [comment]
I think we still need to support older Java versions for the client libraries and adapters. Especially:

- java admin-client
- authz-client
- authz policy-enforcer
- Maybe client-registration client as well (but not 100% sure about this one)

I don't know which is the minimal Java versions, which they support right now, but I afraid that if we change compile version of whole project in parent `pom.xml` to Java 17, then those libraries won't work on earlier java versions than 17 in the runtime?

It will be good to make sure that those are still supported on same Java versions as today IMO.

For SAML adapter, I think the changes are OK in this PR and hopefully we can even increase java version (Based on what is the minimum Java version for Wildfly 28 or EAP 8, added inline comment about it).

File: adapters/spi/jboss-adapter-core/pom.xml
Comment (pskopek) [reply adapters/spi/jboss-adapter-core/pom.xml:34]: Just a note: This needs to stay on JDK 11 as this is lowest version supported by WF/EAP8.

**Comment by pskopek** [comment]
> I think we still need to support older Java versions for the client libraries and adapters. Especially:
> 
> * java admin-client
> * authz-client
> * authz policy-enforcer
> * Maybe client-registration client as well (but not 100% sure about this one)
> 
> I don't know which is the minimal Java versions, which they support right now, but I afraid that if we change compile version of whole project in parent `pom.xml` to Java 17, then those libraries won't work on earlier java versions than 17 in the runtime?
> 
> It will be good to make sure that those are still supported on same Java versions as today IMO.
> 
> For SAML adapter, I think the changes are OK in this PR and hopefully we can even increase java version (Based on what is the minimum Java version for Wildfly 28 or EAP 8, added inline comment about it).

Anything used in WF/EAP8 needs to use Java 11 as this is the lowest version supported by them. 
Quarkus  supports Java 17 and 21

**Comment by jonkoops** [comment]
So which would be the packages still used by Wildfly 28 or EAP 8? I can selectively add Java 11 there so they don't regress. If anyone has a definitive list that would be really helpful.

**Review by pskopek — CHANGES_REQUESTED** [review]
We have to identify which artifacts are used in WF/EAP8 to set their Java version properly. Otherwise the PR looks good to me.

**Review by stianst — DISMISSED** [review]
Anything that is used client-side (integration/admin-client, authz/client, adapters/*) as well as transitive dependencies like common and core, will need to support older Java versions.

Not sure if that still means Java 8, or Java 11.

Personally I think having the root pom having the lowest version, with an explicit override for server-only stuff is the safest and simplest approach, which is what we are currently doing, so would consider thinking a little bit differently about this change. However, I don't feel all that strongly about it, and could go with Java 17 at the root, but we would need to make sure anything used client-side is using the correct version.

Another potential idea to cleanup a bit is to add two properties to the root pom (java.server.version and java.adapter.version), and use those properties throughout so we have a single place to set supported java versions.

File: pom.xml
Comment (stianst) [inline pom.xml:39]: Maven wrapper is using 3.9.6; https://github.com/keycloak/keycloak/blob/main/.mvn/wrapper/maven-wrapper.properties#L17

**Comment by jonkoops** [comment]
I'm putting this PR on hold until #30395 is merged, as it would likely allow me to build on top and get rid of the Java 8 requirement on some pieces of code.

File: pom.xml
Comment (jonkoops) [reply pom.xml:39]: Split this out into a separate PR (see #30502). I'll remove this change after that code is merged in.

Comment (jonkoops) [reply pom.xml]: Done, this version is no longer upgraded. I'll resolve this one.

File: adapters/saml/core/src/main/java/org/keycloak/adapters/saml/profile/ecp/EcpAuthenticationHandler.java
Comment (jonkoops) [inline adapters/saml/core/src/main/java/org/keycloak/adapters/saml/profile/ecp/EcpAuthenticationHandler.java:45]: I was able to move `keycloak-saml-adapter-core` from `javax` to `jakarta`. Now the only code that doesn't compile under Java 17 is `jboss-adapter-core`, but we'll have to use Java 11 there anyways to support EAP 8.

**Comment by jonkoops** [comment]
@pskopek was able to give me a list of the [GAV of EAP 8](https://download.eng.brq.redhat.com/released/JBoss-middleware/eap8/8.0.2/jboss-eap-8.0.2-all-GAVs.txt), the definitive list of all the Keycloak artifacts used are:

```
org.keycloak:galleon-feature-packs-parent
org.keycloak:keycloak-adapter-spi
org.keycloak:keycloak-common
org.keycloak:keycloak-core
org.keycloak:keycloak-crypto-default
org.keycloak:keycloak-crypto-parent
org.keycloak:keycloak-distribution-parent
org.keycloak:keycloak-jboss-adapter-core
org.keycloak:keycloak-parent
org.keycloak:keycloak-saml-adapter-api-public
org.keycloak:keycloak-saml-adapter-core-jakarta
org.keycloak:keycloak-saml-adapter-galleon-pack
org.keycloak:keycloak-saml-core
org.keycloak:keycloak-saml-core-public
org.keycloak:keycloak-saml-undertow-adapter
org.keycloak:keycloak-saml-wildfly-elytron-jakarta-adapter
org.keycloak:keycloak-saml-wildfly-jakarta-subsystem
org.keycloak:keycloak-undertow-adapter-spi
```

Since Java 11 is the minimum version required for EAP 8 we can set that where needed. If I strip out all the artifacts so that only the JARs are left we get:

```diff
-org.keycloak:galleon-feature-packs-parent
org.keycloak:keycloak-adapter-spi
org.keycloak:keycloak-common
org.keycloak:keycloak-core
org.keycloak:keycloak-crypto-default
-org.keycloak:keycloak-crypto-parent
-org.keycloak:keycloak-distribution-parent
org.keycloak:keycloak-jboss-adapter-core
org.keycloak:keycloak-parent
org.keycloak:keycloak-saml-adapter-api-public
org.keycloak:keycloak-saml-adapter-core-jakarta
-org.keycloak:keycloak-saml-adapter-galleon-pack
org.keycloak:keycloak-saml-core
org.keycloak:keycloak-saml-core-public
org.keycloak:keycloak-saml-undertow-adapter
org.keycloak:keycloak-saml-wildfly-elytron-jakarta-adapter
org.keycloak:keycloak-saml-wildfly-jakarta-subsystem
org.keycloak:keycloak-undertow-adapter-spi
```

I've pushed a commit that sets the release version to Java 11 for all these artifacts.

**Comment by jonkoops** [comment]
I've also double checked this list and found the following new dependencies, since the provided GAV is based on Keycloak 22  these were likely not part of that release:

- `org.keycloak:keycloak-server-spi (dependency of `org.keycloak:keycloak-crypto-default`)
- `org.keycloak:keycloak-server-spi-private` (dependency of `org.keycloak:keycloak-crypto-default`)
- `org.keycloak:keycloak-crypto-fips1402` (dependency of `org.keycloak:keycloak-saml-core`, activated by a profile)

**Comment by jonkoops** [comment]
Since I had to lower the target for `keycloak-core` to Java 11, I've [restored the original profile](https://github.com/keycloak/keycloak/compare/41e5fd7982dd11b2499732c8fac58a278b37c80f..4287fc46df7eb091e0f3d93fd38cc74ce90aabcc) to detect Java 16 for some of the JWK code.

File: adapters/spi/jboss-adapter-core/pom.xml
Comment (jonkoops) [reply adapters/spi/jboss-adapter-core/pom.xml:35]: I think we can resolve this thread, as we agree the minimum version will remain 11 until EAP upgrades to a newer version.

File: adapters/saml/core/pom.xml
Comment (jonkoops) [reply adapters/saml/core/pom.xml]: I think we can resolve this thread, as we agree the minimum version will remain 11 until EAP upgrades to a newer version.

**Comment by pskopek** [comment]
> * `org.keycloak:keycloak-server-spi` (dependency of `org.keycloak:keycloak-crypto-default`)
> * `org.keycloak:keycloak-server-spi-private` (dependency of `org.keycloak:keycloak-crypto-default`)
> * `org.keycloak:keycloak-crypto-fips1402` (dependency of `org.keycloak:keycloak-saml-core`, activated by a profile)

The above dependencies are not part of SAML adapter distribution. Can be safely compiled with default Java version (Java 17).

**Comment by jonkoops** [comment]
@pskopek thanks for verifying that, I've removed the Java 11 requirements for those packages.

**Review by pskopek — APPROVED** [review]
LGTM!

**Review by pskopek — APPROVED** [review]

**Review by mposolda — APPROVED** [review]

**Pull request merged**
